### ISSUE TITLE
Please use `Column.base_columns` for `property2field()` conversions.

### ISSUE THREAD

**Issue by mjpieters** [marshmallow-code/marshmallow-sqlalchemy#383]
I've got a more complex model case that triggers a traceback when `property2field()` is trying to determine if a column is auto-incrementing. If I suppress that specific problem, the model doesn't include defaults. Both problems have the same underlying cause: SQLAlchemy columns can be proxies themselves, but getting the original, base column is easy.

Let me demonstrate how I came to that point. The following model definitions are a little bit more involved, in part because having both a 'traditional' model `_Foo` and a sub-query based model `Foo` illustrates the issues nicely, as well as give a real-world motivation for using a subquery-based model:

```python
from uuid import uuid4
from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer
from sqlalchemy.orm import backref, declarative_base, relationship
from sqlalchemy.sql import func, select
from sqlalchemy.dialects.postgresql import UUID

Base = declarative_base()

class BaseMixin:
    id = Column(UUID, primary_key=True, default=uuid4)
    created_at = Column(DateTime, default=func.now())
    updated_at = Column(DateTime, default=func.now(), onupdate=func.now())

class Related(Base, BaseMixin):
    __tablename__ = "related"

class _Foo(Base, BaseMixin):
    __tablename__ = "foo"

    related_id = Column(UUID, ForeignKey("related.id"), nullable=False)
    int_field = Column(Integer, nullable=False, default=1234)
    bool_field = Column(Boolean, nullable=False, default=False)

# Query wrapping the foo table, adding the `number` column. It serves the
# same role as a VIEW, but purely in SQLAlchemy. VIEWs are a beast to manage
# with Alembic / Flask-Migrate and there is currently no advantage in making
# this an actual PostgreSQL view (see, for example, the DBA Stack Exchange
# question discussing view performance at https://dba.stackexchange.com/q/151169)
numbered_foo = (
    select(_Foo)
    .add_columns(
        func.row_number()
        .over(partition_by=_Foo.related_id, order_by=[_Foo.created_at, _Foo.id])
        .label("number")
    )
    .subquery("numbered_foo")
)

class Foo(Base):
    __table__ = numbered_foo

    related = relationship(
        "Related",
        foreign_keys=[numbered_foo.c.related_id],
        backref=backref("foos", order_by=numbered_foo.c.number),
    )
```

The above `Foo` model works perfectly for our needs, where we can produce `Foo` results with sequential numbering per `Related` group. The _Foo model happens to exist because we have existing base mix-ins that it builds upon, and that's just easier with class inheritance.

Let's turn the above into a Marshmallow schema; this includes a schema for `_Foo` so we can compare later on:

```python
from marshmallow.fields import Integer
from marshmallow_sqlalchemy import auto_field, SQLAlchemyAutoSchema

class BaseSchemaMixin:
    id = auto_field(dump_only=True)
    created_at = auto_field(dump_only=True)
    updated_at = auto_field(dump_only=True)

class _FooSchema(BaseSchemaMixin, SQLAlchemyAutoSchema):
    class Meta:
        model = _Foo

class FooSchema(BaseSchemaMixin, SQLAlchemyAutoSchema):
    class Meta:
        model = Foo

    number = Integer(dump_only=True)
```

Note that I don't expect `number` to be inferred here, there is no traditional ORM `Column` object with type information for that one.

Attempting to execute the above results in a traceback:

```python
Traceback (most recent call last):
  File "/.../mcve.py", line 64, in <module>
    class FooSchema(BaseSchemaMixin, SQLAlchemyAutoSchema):
  File "/private/tmp/ms-repro/venv/lib/python3.9/site-packages/marshmallow/schema.py", line 121, in __new__
    klass._declared_fields = mcs.get_declared_fields(
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/schema/sqlalchemy_schema.py", line 91, in get_declared_fields
    fields.update(mcs.get_declared_sqla_fields(fields, converter, opts, dict_cls))
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/schema/sqlalchemy_schema.py", line 130, in get_declared_sqla_fields
    converter.fields_for_model(
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/convert.py", line 141, in fields_for_model
    field = base_fields.get(key) or self.property2field(prop)
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/convert.py", line 183, in property2field
    field_kwargs = self._get_field_kwargs_for_property(prop)
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/convert.py", line 278, in _get_field_kwargs_for_property
    self._add_column_kwargs(kwargs, column)
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/convert.py", line 305, in _add_column_kwargs
    kwargs["required"] = not column.nullable and not _has_default(column)
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/convert.py", line 28, in _has_default
    or _is_auto_increment(column)
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/convert.py", line 33, in _is_auto_increment
    return column.table is not None and column is column.table._autoincrement_column
AttributeError: 'Subquery' object has no attribute '_autoincrement_column'
```

Shorter version without a whole schema:

```python
>>> from marshmallow_sqlalchemy.convert import _is_auto_increment
>>> _is_auto_increment(Foo.__table__.c.id)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../lib/python3.9/site-packages/marshmallow_sqlalchemy/convert.py", line 33, in _is_auto_increment
    return column.table is not None and column is column.table._autoincrement_column
AttributeError: 'Subquery' object has no attribute '_autoincrement_column'
```

Bummer. But before going into what might be wrong, let's add add a work-around for the attribute error:

```python
from functools import wraps
from marshmallow_sqlalchemy import convert

_orig = convert._is_auto_increment

@wraps(_orig)
def safe_is_auto_increment(*args, **kwargs):
    try:
        return _orig(*args, **kwargs)
    except AttributeError:
        return False

convert._is_auto_increment = safe_is_auto_increment
```

and if we then try to create the schemas again, we end up with a `FooSchema` schema with several fields marked as required (`id`, `bool_field`, `int_field`), despite these having default values:

```python
>>> for field in FooSchema().fields.values():
...     print(f"{field.name:<10}  req? {field.required}")
... 
updated_at  req? False
number      req? False
id          req? True
created_at  req? False
bool_field  req? True
int_field   req? True
```

Compare this with `_FooSchema`, the traditional mapping:

```python
>>> for field in _FooSchema().fields.values():
...     print(f"{field.name:<10}  req? {field.required}")
... 
updated_at  req? False
id          req? False
created_at  req? False
bool_field  req? False
int_field   req? False
```

I went digging to see what is going on and found that SQLAlchemy uses an undocumented internal method `._make_proxy()` to essentially wrap each column from the `_Foo` model / `foo` table definition for the `numbered_foo` subquery, and these are then the columns that end up on the `Foo` mapper. The proxies do not copy `nullable` or `default` information across and `table` points to the new From clause.

You can, however, always still access the _original_, base columns via the non-empty [`Column.base_columns` set](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.ColumnElement.base_columns), which are the 'original' columns that each column based is on. For source columns such as those for the `_Foo` model / `foo` table,`.base_columns` is always a set of 1 containing the column itself:

```python
>>> col = _Foo.id.prop.columns[0]
>>> col.base_columns == set([col])
True
```

while for the sub-query columns for `Foo`, that set contains the original column for the `_Foo` mapper:

```python
>>> col = Foo.id.prop.columns[0]
>>> col.base_columns == set([col])
False
>>> [orig] = col.base_columns
>>> orig is _Foo.id.prop.columns[0]
True
```

If I were to replace `prop.columns[0]` with `next(iterable(prop.columns[0].base_columns))` in the Marshmallow-SQLAlchemy field converter, I get much better results:

```python
>>> from marshmallow_sqlalchemy.convert import ModelConverter
>>> conv = ModelConverter()
>>> kw = {}; conv._add_column_kwargs(kw, Foo.__table__.c.id); kw
{'required': True}
>>> kw = {}; conv._add_column_kwargs(kw, next(iter(Foo.__table__.c.id.base_columns))); kw
{'required': False}
```

and the same still works for a traditional, straight-forward model or table definition.

Downside: it you can't use the same path for the `number` column in the demo subquery, but it breaks in almost the same way as before (with or without my proposed change, conversion either fails at the `_get_field_class_for_column` hurdle with `marshmallow_sqlalchemy.exceptions.ModelConversionError:Could not find field column of type <class 'sqlalchemy.sql.sqltypes.NullType'>.`; `Foo.number.prop.columns[0].type` is the `NullType()` here, and if you add in a `class row_number(GenericFunction):` definition with `type = Integer()`, there still is no `nullable` attribute). If you are curious: `Foo.__table__.c.number` is a `<sqlalchemy.sql.elements.ColumnClause` object and `Foo.__table__.c.number.base_columns` contains a single `sqlalchemy.sql.elements.Over` element, so if there is a type known, my proposal results in a subtly different `AttributeError` at this point (it replaces `ColumnClause` with `Over` in the description).

TLDR: please replace `prop.columns[0]` with the contents of `prop.columns[0].base_columns`, or at least when `prop.columns[0]` is not in that set and the set contains a single element.

Versions:

| project | version  |
|:--|:--|
| marshmallow-sqlalchemy | 0.25.0 |
| SQLAlchemy | 1.4.13 |
| marshmallow | 3.11.1 |
| Python | 3.9.2 |

**Comment by mjpieters** [comment]
As a follow-up, for _my specific situation_, the following monkeypatch serves as a work-around:

```python
orig = convert.ModelConverter._add_column_kwargs

@wraps(orig)
def add_base_column_kwargs(self, kwargs, column, *args, **kw):
    if column not in column.base_columns and len(column.base_columns) == 1:
        # This column is a mere proxy for the actual table column, which holds
        # the real metadata needed for proper field property reflection.
        (column,) = column.base_columns
    orig(self, kwargs, column, *args, **kw)

convert.ModelConverter._add_column_kwargs = add_base_column_kwargs
```

but I don't have documentation attached to my columns nor use the (deprecated) `info["marshmallow"]` dictionaries anywhere. If you _do_ use either of those, the above minimal patch won't suffice as `ModelConverter._get_field_kwargs_for_property()` still only replaces `prop` with `prop.column[0]` and will not unwrap further.
